src/mlpack/tests/layer_names_test.cpp

```diff
@@ -0,0 +1,135 @@
+  // Bilinear interpolation is not yet supported by the string converter
+  LayerTypes<> unsupportedLayer = new BilinearInterpolation<>();
+
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      atrousConvolution) == "atrousconvolution");
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      alphaDropout) == "alphadropout");
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      batchNorm) == "batchnorm");
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      constant) == "constant");
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      convolution) == "convolution");
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      dropConnect) == "dropconnect");
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      dropout) == "dropout");
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      flexibleReLU) == "flexiblerelu");
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      layerNorm) == "layernorm");
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      linear) == "linear");
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      linearNoBias) == "linearnobias");
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      maxPooling) == "maxpooling");
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      meanPooling) == "meanpooling");
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      multiplyConstant) == "multiplyconstant");
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      reLULayer) == "relu");
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      transposedConvolution) == "transposedconvolution");
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      identityLayer) == "identity");
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      tanHLayer) == "tanh");
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      eLU) == "elu");
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      hardTanH) == "hardtanh");
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      leakyReLU) == "leakyrelu");
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      pReLU) == "prelu");
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      sigmoidLayer) == "sigmoid");
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      logSoftMax) == "logsoftmax");
+  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+                      unsupportedLayer) == "unsupported");
+  // Delete all instances.
+  delete &atrousConvolution;
```

Comment:
Actually I think that we have to be a little special in how we handle this.  Since `atrousConvolution` and the others are of type `LayerTypes<>` (which is itself a `boost::variant<>` that holds a pointer to layer types), we have to use a visitor to delete them too.  I think this will work:

```
boost::apply_visitor(DeleteVisitor(), atrousConvolution);
```

Reply:
Not sure, but maybe it would be simpler if we go with the other idea? Adding ``static std::string layerName`` to each layer?

Reply:
Not sure. But I think you are correct. I was required to call visitor for `boost::<variant>` while writing tests for `bias visitor`.